jax-stuff.md

```diff
@@ -150,36 +152,86 @@ mesh = jax.make_mesh((4, 2), ('X', 'Y'))
+
+<h3 id="explicit-sharding-mode">Explicit sharding mode</h3>
+
+Auto mode leaves sharding decisions to the compiler. But what if you have a strong opinion about how you want your program sharded? With enough calls to jax.lax.with_sharding_constraint you can probably guide the compiler’s hand to make it do what you want. But “compiler tickling” is famously not a fun programming model. Where should you put the sharding constraints? You could put them on every single intermediate but that’s a lot of work and it’s also easy to make mistakes that way because there’s no way to check that the shardings make sense together. More commonly, people add just enough sharding annotations to constrain the compiler. But this is a slow iterative process. It’s hard to know ahead of time what XLA’s Shardy pass will do (it’s a whole-program optimization) so all you can do is add annotations, inspect XLA’s sharding choices to see what happened by looking at the profile, and repeat.
+
+To fix this we’ve come up with a different style of sharding programming we call “explicit sharding” or “sharding in types”. The idea is that sharding propagation happens at the JAX level at trace time. Each JAX operation has a sharding rule that takes the shardings of the op’s arguments and produces a sharding for the op’s result. For most operations these rules are simple and obvious because there’s only one reasonable choice. But for some operations it’s ambiguous how to shard the result in which case JAX throws a trace-time error and we ask the programmer to provide an `out_sharding` argument explicitly.
+Since the shardings are propagated at trace time they can also be queried at trace time too via `jax.typeof`.
+
+```
+import jax
+import jax.numpy as jnp
+import jax.sharding as shd
+
+# Running on an TPU v5e 2x2. This assigns names to the two physical axes of the hardware.
+mesh = jax.make_mesh(axis_shapes=(2, 2), axis_names=('X', 'Y'),
+                                       axis_types=(shd.AxisType.Explicit, shd.AxisType.Explicit))
+
+# This tells JAX to use this mesh for all operations, so you can just specify the PartitionSpec P.
+jax.set_mesh(mesh)
+
+# We create a matrix W and input activations In sharded across our devices.
+In = jnp.zeros((8, 2048), dtype=jnp.bfloat16, out_sharding=jax.P('X', 'Y'))
+W = jnp.zeros((2048, 8192), dtype=jnp.bfloat16, out_sharding=jax.P('Y', None))
+
+@jax.jit
+def matmul_square(In, W):
+  print(jax.typeof(In))  # bfloat16[8@X, 2048@Y]
+  print(jax.typeof(W))  # bfloat16[2048@Y, 8192]
+  return jnp.einsum('bd,df->bf', jnp.square(In), W)
+
+matmul_square(In, W)  # This will error
+```
+
+This code errors with `Contracting dimensions are sharded and it is ambiguous how the output should be sharded. Please specify the output sharding via the `out_sharding` parameter. Got lhs_contracting_spec=('Y',) and rhs_contracting_spec=('Y',)`
```

Comment:
Could we provide a working example first, describe how it works, then provide a failing example? That way people can have a working code example first?

Reply:
I don't agree. I think SHIT has two parts: first you show that you can see the sharding in types and that it's done a simple thing correctly, and then you show an example with ambiguity where it asks for clarification.

Reply:
Done